cli/2024-06-04-aws-irsa-support.md

```diff
@@ -0,0 +1,326 @@
+  IRSA,
+}
+```
+
+```
+model AwsAccessKeyCredentialProperties extends AwsCredentialProperties {
+  @doc("IRSA kind")
+  kind: AWSCredentialKind.IRSA;
+ 
+  @doc("The roleARN")
+  roleARN: string;
+
+  @doc("The storage properties")
+  storage: CredentialStorageProperties;
+}
+```
+
+Based on this new model we have to update API calls. 
+
+## User Experience/ CLI Design
+
+`rad init` and `rad credential` commands should be updated to support AWS IRSA.
+Also, `rad install kubernetes --set` should be able to set a AWS role-arn. 
+
+### non interactive support
+rad install kubernetes --set global.aws.IRSA.enabled=true #default false
+
+if enabled, it should print a success/ failure message after updating the pod spec to mount the token. It should also print a information message to the complete of configuration by using rad credential register and rad env update.
+
+rad credential register aws --type <IRSA> --role <roleARN>
+rad credential register aws --type <access-key> --access-key-id <access-key-id> --secret-access-key <secret-access-key>
+
+rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 
+
+### interactive rad init support
+```
+% rad init --full  
+Enter an environment name 
+>default 
+
+% rad init --full  
+Select your cloud provider                     
+1. Azure                                  
+> 2. AWS                                     
+3.[back] 
+
+% rad init --full  
+
+Select the identity for the AWS cloud provider 
+1. access key                              
+> 2. IRSA  
+
+% rad init --full  
+
+Enter the AWS RoleARN for configuring IRSA: <role-arn-radius>
+
+```
+
+
+Once radius is installed, we use rad env update to update the aws provider config
+```
+% rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 
+```
+
+**Sample Output:**
+
+n/a
+
+**Sample Recipe Contract:**
+
+n/a
+
+### CLI Design
+
+`rad init --full` will need to be updated to ask the user to choose between AWS access keys and IRSA. rad credential register aws should now support two credential types, and associated parameters.
+
+We should update the Radius Helm chart to allow the user to enable IRSA during installation with the `rad install kubernets --set global.aws.IRSA.enabled=true` value. This must be passed to ucp and rp charts through values and when the value is set the pod spec should mount the service-account token.
+
+### Multi-tenancy
+
+Currently, we support only one credential. However, for future, we might want to support multiple credentials potentially across AWS accounts ( say, one for QA team and other for Dev). The code changes for IRSA should be compatible with these changes in future. Since IRSA setup establishes trust between the cluster - oidc provider and AWS, once we support multiple credentials, we should be able to assume different roleARN based on which environment we are deploying from. We just have to establish right trust entities in all roles.  
```

Comment:
Is this implying that multi-tenancy is out of scope for this design because Radius doesn't support multiple credentials in general today?

Reply:
I think the code implementation for this IRSA support still can take care of future multi-tenancy though it is not supported today. @nithyatsu to more on this.

Reply:
yes, like @shalabhms  mentions, the proposed IRSA implementation can take care of multi-tenancy in future(when we support multiple credentials)

Reply:
add a statement "when radius supports multi-tenancy, IRSA will be compatible with it".